File: src/hypeedge/market_data/live_provider.py
"""Live market data provider combining WS + REST + BookManager (Phase 1).

Implements the MarketDataProvider Protocol defined in provider.py.
Subscribes to EventBus events to maintain in-memory trade state,
delegates book queries to BookManager, and backfill to RestClient.
"""

from __future__ import annotations

import asyncio
import contextlib
from datetime import UTC, datetime
from typing import Any

import structlog

from hypeedge.config.settings import AppSettings
from hypeedge.core.events import (
    EVENT_CANDLE_UPDATE,
    EVENT_FUNDING_UPDATE,
    EVENT_MID_PRICE_UPDATE,
    EVENT_TRADE_UPDATE,
    Event,
    EventBus,
)
from hypeedge.core.models import Candle, FundingRate, L2BookSnapshot, Trade
from hypeedge.core.types import Symbol, Timestamp
from hypeedge.market_data.book import BookManager
from hypeedge.market_data.provider import MarketPriceSnapshot
from hypeedge.market_data.rest_client import RestClient

logger = structlog.get_logger(__name__)


class LiveMarketDataProvider:
    """Concrete MarketDataProvider backed by live WS feed + REST backfill.

    Provides synchronous (zero-latency) access to current market state
    for strategies, and async methods for historical data retrieval.
    """

    def __init__(
        self,
        settings: AppSettings,
        event_bus: EventBus,
        rest_client: RestClient,
        book_manager: BookManager,
    ) -> None:
        self._settings = settings
        self._event_bus = event_bus
        self._rest_client = rest_client
        self._book_manager = book_manager

        # In-memory latest state per symbol
        self._last_trades: dict[Symbol, Trade] = {}
        self._mid_prices: dict[Symbol, float] = {}
        self._mid_price_timestamps: dict[Symbol, datetime] = {}
        self._mid_price_versions: dict[Symbol, int] = {}
        self._mid_price_connection_generations: dict[Symbol, int] = {}
        self._mid_price_exchange_timestamps: dict[Symbol, Timestamp | None] = {}
        self._funding: dict[Symbol, FundingRate] = {}
        self._candles: dict[tuple[Symbol, str], list[Candle]] = {}
        self._candle_backfill_locks: dict[tuple[Symbol, str], asyncio.Lock] = {}
        self._max_candles_per_series = 1_500

        self._running = False
        self._subscriber_task: asyncio.Task[None] | None = None
        self._subscriptions: list[tuple[str, asyncio.Queue[Event]]] = []

    async def start(self) -> None:
        """Subscribe to events and begin tracking market state."""
        self._running = True

        self._subscriptions = [
            (EVENT_TRADE_UPDATE, self._event_bus.subscribe(EVENT_TRADE_UPDATE)),
            (EVENT_MID_PRICE_UPDATE, self._event_bus.subscribe(EVENT_MID_PRICE_UPDATE)),
            (EVENT_FUNDING_UPDATE, self._event_bus.subscribe(EVENT_FUNDING_UPDATE)),
            (EVENT_CANDLE_UPDATE, self._event_bus.subscribe(EVENT_CANDLE_UPDATE)),
        ]

        self._subscriber_task = asyncio.create_task(
            self._consume_events([queue for _, queue in self._subscriptions]),
            name="market_data_provider",
        )
        logger.info("live_market_data_provider_started")

    async def stop(self) -> None:
        """Stop consuming events."""
        self._running = False
        if self._subscriber_task:
            self._subscriber_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._subscriber_task
            self._subscriber_task = None
        for event_type, queue in self._subscriptions:
            self._event_bus.unsubscribe(event_type, queue)
        self._subscriptions.clear()
        logger.info("live_market_data_provider_stopped")

    # --- MarketDataProvider interface ---

    def get_book(self, symbol: Symbol) -> L2BookSnapshot | None:
        """Get the latest L2 order book snapshot."""
        return self._book_manager.get_snapshot(symbol)

    def get_mid_price(self, symbol: Symbol) -> float | None:
        """Get the latest mid price for a symbol.

        Prefers the allMids WS price (most accurate). Falls back to
        (best_bid + best_ask) / 2 from the order book.
        """
        mid = self._mid_prices.get(symbol)
        if mid is not None:
            return mid
        return self._book_manager.get_mid_price(symbol)

    def get_price_snapshot(self, symbol: Symbol) -> MarketPriceSnapshot | None:
        mid = self._mid_prices.get(symbol)
        observed_at = self._mid_price_timestamps.get(symbol)
        if mid is not None and observed_at is not None:
            return MarketPriceSnapshot(
                mid,
                observed_at,
                exchange_ts=self._mid_price_exchange_timestamps.get(symbol),
                version=self._mid_price_versions.get(symbol, 0),
                connection_generation=self._mid_price_connection_generations.get(symbol, 0),
            )
        book = self._book_manager.get_snapshot(symbol)
        book_mid = self._book_manager.get_mid_price(symbol)
        if book is None or book_mid is None:
            return None
        return MarketPriceSnapshot(
            book_mid,
            book.received_at,
            exchange_ts=book.exchange_ts,
            version=book.version,
            connection_generation=book.connection_generation,
        )

    def get_last_trade(self, symbol: Symbol) -> Trade | None:
        """Get the most recent trade for a symbol."""
        return self._last_trades.get(symbol)

    def get_funding(self, symbol: Symbol) -> FundingRate | None:
        """Get the latest normalized funding/mark/open-interest snapshot."""
        return self._funding.get(symbol)

    def get_candles(self, symbol: Symbol, interval: str, limit: int = 300) -> list[Candle]:
        """Get recent candles in ascending timestamp order."""
        series = self._candles.get((symbol, interval), [])
        return list(series[-limit:])

    async def ensure_candles(
        self,
        symbol: Symbol,
        interval: str,
        limit: int,
        start_ms: int,
        end_ms: int,
    ) -> list[Candle]:
        """Warm a candle series once, coalescing concurrent API requests."""
        key = (symbol, interval)
        lock = self._candle_backfill_locks.setdefault(key, asyncio.Lock())
        async with lock:
            cached = self.get_candles(symbol, interval, limit)
            if len(cached) >= limit:
                return cached
            history = await self.backfill_candles(symbol, interval, start_ms, end_ms)
            for candle in history:
                self._handle_candle(candle)
            return self.get_candles(symbol, interval, limit)

    async def backfill_candles(self, symbol: Symbol, interval: str, start_ms: int, end_ms: int) -> list[Candle]:
        """Backfill historical candle data via REST."""
        return await self._rest_client.backfill_candles(
            coin=str(symbol),
            interval=interval,
            start_time=start_ms,
            end_time=end_ms,
        )

    async def backfill_funding(self, symbol: Symbol, start_ms: int, end_ms: int) -> list[FundingRate]:
        """Backfill historical funding rate data via REST."""
        return await self._rest_client.backfill_funding(
            coin=str(symbol),
            start_time=start_ms,
            end_time=end_ms,
        )

    # --- Internal event consumption ---

    async def _consume_events(
        self,
        queues: list[asyncio.Queue[Event]],
    ) -> None:
        """Consume trade and mid-price events to maintain latest state."""
        readers = {
            asyncio.create_task(queue.get(), name=f"provider_reader_{index}"): queue
            for index, queue in enumerate(queues)
        }

        try:
            while self._running:
                done, _ = await asyncio.wait(
                    readers,
                    timeout=1.0,
                    return_when=asyncio.FIRST_COMPLETED,
                )

                for task in done:
                    queue = readers.pop(task)
                    try:
                        event: Event = task.result()
                        self._handle_event(event)
                    except asyncio.CancelledError:
                        raise
                    except Exception:
                        logger.exception("provider_event_error")

                    readers[asyncio.create_task(queue.get(), name=task.get_name())] = queue

        except asyncio.CancelledError:
            logger.debug("live_provider_cancelled")
        finally:
            for task in readers:
                task.cancel()
            await asyncio.gather(*readers, return_exceptions=True)

    def _handle_event(self, event: Event) -> None:
        """Route an event to the appropriate handler."""
        if event.event_type == EVENT_TRADE_UPDATE:
            self._handle_trade(event.payload)
        elif event.event_type == EVENT_MID_PRICE_UPDATE:
            self._handle_mid_price(event.payload, event.timestamp)
        elif event.event_type == EVENT_FUNDING_UPDATE:
            self._handle_funding(event.payload)
        elif event.event_type == EVENT_CANDLE_UPDATE:
            self._handle_candle(event.payload)

    def _handle_trade(self, payload: Any) -> None:
        """Update latest trade state."""
        if isinstance(payload, Trade):
            self._last_trades[payload.symbol] = payload

    def _handle_mid_price(self, payload: Any, event_received_at: datetime | None = None) -> None:
        """Update mid price from allMids."""
        if isinstance(payload, dict):
            symbol = payload.get("symbol")
            price = payload.get("price")
            if symbol is not None and price is not None:
                self._mid_prices[symbol] = float(price)
                received_at = payload.get("received_at", event_received_at)
                self._mid_price_timestamps[symbol] = (
                    received_at if isinstance(received_at, datetime) else datetime.now(UTC)
                )
                self._mid_price_versions[symbol] = self._mid_price_versions.get(symbol, 0) + 1
                generation = payload.get("connection_generation", 0)
                self._mid_price_connection_generations[symbol] = generation if isinstance(generation, int) else 0
                exchange_ts = payload.get("exchange_ts")
                self._mid_price_exchange_timestamps[symbol] = (
                    Timestamp(int(exchange_ts)) if isinstance(exchange_ts, (int, float)) else None
                )

    def _handle_funding(self, payload: Any) -> None:
        if isinstance(payload, FundingRate):
            self._funding[payload.symbol] = payload

    def _handle_candle(self, payload: Any) -> None:
        if not isinstance(payload, Candle):
            return
        key = (payload.symbol, payload.interval)
        series = self._candles.setdefault(key, [])
        if series and series[-1].timestamp == payload.timestamp:
            series[-1] = payload
        elif not series or payload.timestamp > series[-1].timestamp:
            series.append(payload)
        else:
            by_timestamp = {candle.timestamp: candle for candle in series}
            by_timestamp[payload.timestamp] = payload
            series[:] = sorted(by_timestamp.values(), key=lambda candle: candle.timestamp)
        if len(series) > self._max_candles_per_series:
            del series[: len(series) - self._max_candles_per_series]

File: crates/trading/src/market_data/live_provider.rs
//! Live market data provider, port of
//! `src/hypeedge/market_data/live_provider.py`.
//!
//! Implements the domain [`MarketDataProvider`] facade by combining the WS
//! feed (via the event bus), the shared [`BookManager`], and a REST history
//! client for candle/funding backfill. Latest trade/mid/funding state is
//! maintained by consuming the bus in a background task.

use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use hypeedge_domain::decimal::Decimal;
use hypeedge_domain::error::HypeEdgeError;
use hypeedge_domain::events::{DomainEvent, Event, EventType};
use hypeedge_domain::models::{Candle, FundingRate, L2BookSnapshot, MidPrice, Trade};
use hypeedge_domain::traits::MarketDataProvider;
use hypeedge_infra::event_bus::{BoundedMailbox, EventBus};

/// A subscriber mailbox (shared handle to a bounded event queue).
type Mailbox = Arc<BoundedMailbox<Arc<Event>>>;

use super::book::BookManager;

/// A normalized mid/mark price and its actual observation time.
#[derive(Debug, Clone, PartialEq)]
pub struct MarketPriceSnapshot {
    pub price: f64,
    pub observed_at: DateTime<Utc>,
    pub exchange_ts: Option<i64>,
    pub version: u64,
    pub connection_generation: u32,
}

/// The REST history boundary used for backfill (implemented by the REST
/// client / app wiring).
#[async_trait]
pub trait CandleHistoryClient: Send + Sync {
    async fn backfill_candles(
        &self,
        coin: &str,
        interval: &str,
        start_time: i64,
        end_time: i64,
    ) -> Result<Vec<Candle>, HypeEdgeError>;

    async fn backfill_funding(
        &self,
        coin: &str,
        start_time: i64,
        end_time: i64,
    ) -> Result<Vec<FundingRate>, HypeEdgeError>;
}

/// Concrete [`MarketDataProvider`] backed by live WS events + REST backfill.
pub struct LiveMarketDataProvider {
    event_bus: Arc<EventBus>,
    rest_client: Arc<dyn CandleHistoryClient>,
    book_manager: Arc<tokio::sync::Mutex<BookManager>>,
    max_candles_per_series: usize,

    last_trades: tokio::sync::Mutex<HashMap<String, Trade>>,
    mid_prices: tokio::sync::Mutex<HashMap<String, MidPriceState>>,
    funding: tokio::sync::Mutex<HashMap<String, FundingRate>>,
    candles: tokio::sync::Mutex<HashMap<(String, String), Vec<Candle>>>,
}

/// Per-symbol mid-price bookkeeping (mirrors the Python dicts).
#[derive(Debug, Clone)]
struct MidPriceState {
    price: f64,
    observed_at: DateTime<Utc>,
    version: u64,
    connection_generation: u32,
    exchange_ts: Option<i64>,
}

impl LiveMarketDataProvider {
    pub fn new(
        event_bus: Arc<EventBus>,
        rest_client: Arc<dyn CandleHistoryClient>,
        book_manager: Arc<tokio::sync::Mutex<BookManager>>,
    ) -> Self {
        Self {
            event_bus,
            rest_client,
            book_manager,
            max_candles_per_series: 1_500,
            last_trades: tokio::sync::Mutex::new(HashMap::new()),
            mid_prices: tokio::sync::Mutex::new(HashMap::new()),
            funding: tokio::sync::Mutex::new(HashMap::new()),
            candles: tokio::sync::Mutex::new(HashMap::new()),
        }
    }

    /// Subscribe to events and begin tracking market state in a background task.
    pub fn start(self: &Arc<Self>) -> Mailbox {
        let mailbox = self.event_bus.subscribe_many(&[
            EventType::TradeUpdate,
            EventType::MidPriceUpdate,
            EventType::FundingUpdate,
            EventType::CandleUpdate,
            EventType::L2BookUpdate,
        ]);
        let consumer = mailbox.clone();
        let provider = self.clone();
        tokio::spawn(async move {
            provider.consume_events(consumer).await;
        });
        tracing::info!("live_market_data_provider_started");
        mailbox
    }

    /// The consumer loop; run in the background task.
    async fn consume_events(&self, mailbox: Mailbox) {
        while let Some(event) = mailbox.recv().await {
            self.handle_event(&event).await;
        }
    }

    async fn handle_event(&self, event: &Event) {
        match &event.payload {
            // 6c: bridge WS l2Book snapshots into the shared book manager the
            // API and execution engine read.
            DomainEvent::L2BookUpdate(book) => {
                let mut books = self.book_manager.lock().await;
                books.apply_snapshot(book);
            }
            DomainEvent::TradeUpdate(trade) => {
                self.last_trades
                    .lock()
                    .await
                    .insert(trade.symbol.clone(), trade.clone());
            }
            DomainEvent::MidPriceUpdate(mid) => {
                let mut mid_prices = self.mid_prices.lock().await;
                let current = mid_prices.get(&mid.symbol).cloned();
                let version = current.as_ref().map(|c| c.version + 1).unwrap_or(1);
                mid_prices.insert(
                    mid.symbol.clone(),
                    MidPriceState {
                        price: mid.price.to_string().parse::<f64>().unwrap_or(0.0),
                        observed_at: Utc::now(),
                        version,
                        connection_generation: current
                            .as_ref()
                            .map(|c| c.connection_generation)
                            .unwrap_or(0),
                        exchange_ts: Some(mid.timestamp),
                    },
                );
            }
            DomainEvent::FundingUpdate(funding) => {
                let mut funding_map = self.funding.lock().await;
                let current = funding_map.get(&funding.symbol).cloned();
                // Ignore out-of-order or lower-quality snapshots.
                if let Some(current) = &current {
                    if funding.timestamp < current.timestamp {
                        return;
                    }
                    if funding.timestamp == current.timestamp
                        && current.mark_price.inner() > Decimal::ZERO
                        && funding.mark_price.inner() <= Decimal::ZERO
                    {
                        return;
                    }
                }
                funding_map.insert(funding.symbol.clone(), funding.clone());
            }
            DomainEvent::CandleUpdate(candle) => {
                let mut candles = self.candles.lock().await;
                let series = candles
                    .entry((candle.symbol.clone(), candle.interval.clone()))
                    .or_default();
                LiveMarketDataProvider::upsert_candle(series, candle.clone(), self.max_candles_per_series);
            }
            _ => {}
        }
    }

    /// Insert/replace a candle keeping the series time-ordered and bounded.
    fn upsert_candle(series: &mut Vec<Candle>, candle: Candle, max: usize) {
        if let Some(last) = series.last_mut()
            && last.timestamp == candle.timestamp
        {
            *last = candle;
            return;
        }
        if series.is_empty() || series.last().is_some_and(|c| candle.timestamp > c.timestamp) {
            series.push(candle);
        } else {
            let mut by_ts: HashMap<i64, Candle> = series.iter().map(|c| (c.timestamp, c.clone())).collect();
            by_ts.insert(candle.timestamp, candle);
            let mut merged: Vec<Candle> = by_ts.into_values().collect();
            merged.sort_by_key(|c| c.timestamp);
            *series = merged;
        }
        if series.len() > max {
            series.drain(..series.len() - max);
        }
    }

    /// Get the latest order book snapshot (perp or spot share the BookManager).
    pub async fn get_book(&self, symbol: &str) -> Option<L2BookSnapshot> {
        self.book_manager.lock().await.get_snapshot(symbol)
    }

    /// Get the latest mid price; prefers the allMids WS price, falls back to
    /// the book mid.
    pub async fn get_mid_price(&self, symbol: &str) -> Option<f64> {
        let mid_prices = self.mid_prices.lock().await;
        if let Some(state) = mid_prices.get(symbol) {
            return Some(state.price);
        }
        drop(mid_prices);
        self.book_manager.lock().await.get_mid_price(symbol)
    }

    /// Return a normalized mid/mark price and its actual observation time.
    pub async fn get_price_snapshot_full(&self, symbol: &str) -> Option<MarketPriceSnapshot> {
        let mid_prices = self.mid_prices.lock().await;
        if let Some(state) = mid_prices.get(symbol) {
            return Some(MarketPriceSnapshot {
                price: state.price,
                observed_at: state.observed_at,
                exchange_ts: state.exchange_ts,
                version: state.version,
                connection_generation: state.connection_generation,
            });
        }
        drop(mid_prices);
        let book = self.book_manager.lock().await;
        let snapshot = book.get_snapshot(symbol)?;
        let mid = book.get_mid_price(symbol)?;
        Some(MarketPriceSnapshot {
            price: mid,
            observed_at: snapshot.local_ts,
            exchange_ts: Some(snapshot.timestamp),
            version: snapshot.version,
            connection_generation: snapshot.connection_generation,
        })
    }

    pub async fn get_last_trade(&self, symbol: &str) -> Option<Trade> {
        self.last_trades.lock().await.get(symbol).cloned()
    }

    pub async fn get_funding(&self, symbol: &str) -> Option<FundingRate> {
        self.funding.lock().await.get(symbol).cloned()
    }

    pub async fn get_candles(&self, symbol: &str, interval: &str, limit: usize) -> Vec<Candle> {
        let candles = self.candles.lock().await;
        let series = candles.get(&(symbol.to_string(), interval.to_string()));
        let Some(series) = series else { return Vec::new() };
        let start = series.len().saturating_sub(limit);
        series[start..].to_vec()
    }

    /// Warm a candle series once, coalescing concurrent API requests.
    pub async fn ensure_candles(
        &self,
        symbol: &str,
        interval: &str,
        limit: usize,
        start_ms: i64,
        end_ms: i64,
    ) -> Result<Vec<Candle>, HypeEdgeError> {
        let cached = self.get_candles(symbol, interval, limit).await;
        if cached.len() >= limit {
            return Ok(cached);
        }
        let history = self
            .rest_client
            .backfill_candles(symbol, interval, start_ms, end_ms)
            .await?;
        let mut candles = self.candles.lock().await;
        let series = candles
            .entry((symbol.to_string(), interval.to_string()))
            .or_default();
        for candle in history {
            LiveMarketDataProvider::upsert_candle(series, candle, self.max_candles_per_series);
        }
        Ok(self
            .get_candles(symbol, interval, limit)
            .await)
    }

    pub async fn backfill_candles(
        &self,
        symbol: &str,
        interval: &str,
        start_ms: i64,
        end_ms: i64,
    ) -> Result<Vec<Candle>, HypeEdgeError> {
        self.rest_client
            .backfill_candles(symbol, interval, start_ms, end_ms)
            .await
    }

    pub async fn backfill_funding(
        &self,
        symbol: &str,
        start_ms: i64,
        end_ms: i64,
    ) -> Result<Vec<FundingRate>, HypeEdgeError> {
        self.rest_client
            .backfill_funding(symbol, start_ms, end_ms)
            .await
    }
}

#[async_trait]
impl MarketDataProvider for LiveMarketDataProvider {
    async fn get_price_snapshot(
        &self,
        symbol: &str,
    ) -> Result<Option<MidPrice>, HypeEdgeError> {
        let snapshot = self.get_price_snapshot_full(symbol).await;
        Ok(snapshot.map(|s| MidPrice {
            symbol: symbol.to_string(),
            price: Decimal::from_f64(s.price).unwrap_or_default(),
            timestamp: s.exchange_ts.unwrap_or_else(|| Utc::now().timestamp_millis()),
        }))
    }

    async fn get_best_bid_ask(
        &self,
        symbol: &str,
    ) -> Result<Option<(Decimal, Decimal)>, HypeEdgeError> {
        let book = self.get_book(symbol).await;
        let Some(book) = book else { return Ok(None) };
        let Some(bid) = book.bids.first() else { return Ok(None) };
        let Some(ask) = book.asks.first() else { return Ok(None) };
        Ok(Some((bid.price.inner(), ask.price.inner())))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use hypeedge_domain::decimal::{Price, Size};
    use hypeedge_domain::enums::Side;
    use hypeedge_domain::events::EventType;
    use hypeedge_domain::models::L2Level;

    struct StubHistory;

    #[async_trait]
    impl CandleHistoryClient for StubHistory {
        async fn backfill_candles(
            &self,
            _coin: &str,
            _interval: &str,
            _start_time: i64,
            _end_time: i64,
        ) -> Result<Vec<Candle>, HypeEdgeError> {
            Ok(vec![])
        }
        async fn backfill_funding(
            &self,
            _coin: &str,
            _start_time: i64,
            _end_time: i64,
        ) -> Result<Vec<FundingRate>, HypeEdgeError> {
            Ok(vec![])
        }
    }

    fn candle(symbol: &str, interval: &str, ts: i64) -> Candle {
        Candle {
            symbol: symbol.into(),
            interval: interval.into(),
            open: Price::new(Decimal::ONE),
            high: Price::new(Decimal::ONE),
            low: Price::new(Decimal::ONE),
            close: Price::new(Decimal::ONE),
            volume: Size::new(Decimal::ONE),
            timestamp: ts,
        }
    }

    fn provider() -> LiveMarketDataProvider {
        let bus = Arc::new(EventBus::new(1024));
        let history: Arc<dyn CandleHistoryClient> = Arc::new(StubHistory);
        let books = Arc::new(tokio::sync::Mutex::new(BookManager::new(10)));
        LiveMarketDataProvider::new(bus, history, books)
    }

    #[tokio::test]
    async fn upsert_candle_replaces_same_timestamp() {
        let mut series = vec![candle("BTC", "1m", 100)];
        LiveMarketDataProvider::upsert_candle(&mut series, candle("BTC", "1m", 100), 10);
        assert_eq!(series.len(), 1);
    }

    #[tokio::test]
    async fn upsert_candle_appends_and_bounds() {
        let mut series = Vec::new();
        for ts in 0..5 {
            LiveMarketDataProvider::upsert_candle(&mut series, candle("BTC", "1m", ts), 3);
        }
        assert_eq!(series.len(), 3);
        assert_eq!(series[0].timestamp, 2);
    }

    #[tokio::test]
    async fn upsert_candle_sorts_out_of_order() {
        let mut series = Vec::new();
        LiveMarketDataProvider::upsert_candle(&mut series, candle("BTC", "1m", 100), 10);
        LiveMarketDataProvider::upsert_candle(&mut series, candle("BTC", "1m", 50), 10);
        assert_eq!(series.len(), 2);
        assert_eq!(series[0].timestamp, 50);
        assert_eq!(series[1].timestamp, 100);
    }

    #[tokio::test]
    async fn handles_funding_out_of_order_and_lower_quality() {
        let p = provider();
        let ts = Utc::now().timestamp_millis();
        let f1 = FundingRate {
            symbol: "BTC".into(),
            funding_rate: 0.1,
            premium: 0.1,
            mark_price: Price::new(Decimal::from_scaled(50000, 0)),
            open_interest: 0.0,
            timestamp: ts,
        };
        let stale = FundingRate {
            symbol: "BTC".into(),
            funding_rate: 0.1,
            premium: 0.1,
            mark_price: Price::new(Decimal::from_scaled(50000, 0)),
            open_interest: 0.0,
            timestamp: ts - 1,
        };
        p.handle_event(&Event::new(DomainEvent::FundingUpdate(f1.clone())))
            .await;
        p.handle_event(&Event::new(DomainEvent::FundingUpdate(stale.clone())))
            .await;
        // Stale (older timestamp) must be ignored.
        let funding = p.get_funding("BTC").await.unwrap();
        assert_eq!(funding.timestamp, ts);
    }

    #[tokio::test]
    async fn tracks_last_trade_and_mid() {
        let p = provider();
        let trade = Trade {
            symbol: "BTC".into(),
            price: Price::new(Decimal::from_scaled(50000, 0)),
            size: Size::new(Decimal::ONE),
            side: Side::Buy,
            tid: 1,
            timestamp: 100,
            local_ts: Utc::now(),
        };
        p.handle_event(&Event::new(DomainEvent::TradeUpdate(trade.clone())))
            .await;
        assert_eq!(p.get_last_trade("BTC").await.unwrap().tid, 1);

        let mid = MidPrice {
            symbol: "BTC".into(),
            price: Decimal::from_scaled(50010, 0),
            timestamp: 200,
        };
        p.handle_event(&Event::new(DomainEvent::MidPriceUpdate(mid)))
            .await;
        assert_eq!(p.get_mid_price("BTC").await, Some(50010.0));
    }

    #[tokio::test]
    async fn candle_upsert_bounds_series() {
        let mut series = Vec::new();
        for ts in 0..2000 {
            LiveMarketDataProvider::upsert_candle(&mut series, candle("ETH", "5m", ts), 1500);
        }
        assert_eq!(series.len(), 1500);
        assert_eq!(series.first().unwrap().timestamp, 500);
    }

    #[test]
    fn event_types_supported() {
        // Compile-time: these event types exist and are consumed.
        let _ = EventType::TradeUpdate;
        let _ = EventType::MidPriceUpdate;
        let _ = EventType::FundingUpdate;
        let _ = EventType::CandleUpdate;
    }

    #[tokio::test]
    async fn l2_book_update_bridges_into_shared_book() {
        // 6c: an L2BookUpdate on the bus must write the shared BookManager the
        // API and execution engine read.
        let bus = Arc::new(EventBus::new(16));
        let history: Arc<dyn CandleHistoryClient> = Arc::new(StubHistory);
        let books = Arc::new(tokio::sync::Mutex::new(BookManager::new(10)));
        let p = LiveMarketDataProvider::new(bus, history, books.clone());

        let snapshot = L2BookSnapshot {
            symbol: "BTC".into(),
            bids: vec![L2Level {
                price: Price::new(Decimal::from_str_lenient("99").unwrap()),
                size: Size::new(Decimal::from_str_lenient("2").unwrap()),
            }],
            asks: vec![L2Level {
                price: Price::new(Decimal::from_str_lenient("101").unwrap()),
                size: Size::new(Decimal::from_str_lenient("3").unwrap()),
            }],
            timestamp: 1_700_000_000_000,
            local_ts: Utc::now(),
            version: 1,
            connection_generation: 1,
        };
        p.handle_event(&Event::new(DomainEvent::L2BookUpdate(snapshot)))
            .await;

        let shared = books.lock().await.get_snapshot("BTC").expect("book populated");
        assert_eq!(shared.bids.len(), 1);
        assert_eq!(shared.bids[0].price.to_string(), "99");
        assert_eq!(shared.asks[0].price.to_string(), "101");
    }
}
